File: app/src/main/java/org/literacyapp/keyboard/ImeService.java
package org.literacyapp.keyboard;

import android.inputmethodservice.InputMethodService;
import android.inputmethodservice.Keyboard;
import android.inputmethodservice.KeyboardView;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.util.Log;
import android.view.KeyEvent;
import android.view.View;
import android.view.inputmethod.InputConnection;

import java.util.Locale;

public class ImeService extends InputMethodService implements KeyboardView.OnKeyboardActionListener {

    private KeyboardView keyboardView;

    private Keyboard keyboard;

    private boolean caps = false;

    @Override
    public void onCreate() {
        Log.i(getClass().getName(), "onCreate");
        super.onCreate();
    }

    @Override
    public View onCreateInputView() {
        Log.i(getClass().getName(), "onCreateInputView");

        keyboardView = (KeyboardView) getLayoutInflater().inflate(R.layout.keyboard, null);
        keyboard = new Keyboard(this, R.xml.qwerty);
        keyboardView.setKeyboard(keyboard);
        keyboardView.setPreviewEnabled(false);
        keyboardView.setOnKeyboardActionListener(this);

        return keyboardView;
    }

    @Override
    public void updateFullscreenMode() {
        Log.i(getClass().getName(), "updateFullscreenMode");
//        super.updateFullscreenMode();
    }

    @Override
    public void onPress(int i) {

    }

    @Override
    public void onRelease(int i) {

    }

    @Override
    public void onKey(int primaryCode, int[] keyCodes) {
        InputConnection ic = getCurrentInputConnection();

        Locale locale = getResources().getConfiguration().locale;
        if (locale.getLanguage().startsWith("sw")) {
            playClickSw(primaryCode);
        } else {
            playClick(primaryCode);
        }

        switch(primaryCode){
            case Keyboard.KEYCODE_DELETE:
                ic.deleteSurroundingText(1, 0);
                break;
            case Keyboard.KEYCODE_SHIFT:
                caps = !caps;
                keyboard.setShifted(caps);
                keyboardView.invalidateAllKeys();
                break;
            case Keyboard.KEYCODE_DONE:
                ic.sendKeyEvent(new KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_ENTER));
                break;
            default:
                char code = (char)primaryCode;
                if (Character.isLetter(code) && caps) {
                    code = Character.toUpperCase(code);
                }
                ic.commitText(String.valueOf(code), 1);
        }
    }

    private void playClick(int keyCode){
        AudioManager audioManager = (AudioManager) getSystemService(AUDIO_SERVICE);
        MediaPlayer mediaPlayer = null;
        switch(keyCode) {
            case 48:
                mediaPlayer = MediaPlayer.create(this, R.raw.digit_0);
                mediaPlayer.start();
                break;
            case 49:
                mediaPlayer = MediaPlayer.create(this, R.raw.digit_1);
                mediaPlayer.start();
                break;
            case 50:
                mediaPlayer = MediaPlayer.create(this, R.raw.digit_2);
                mediaPlayer.start();
                break;
            case 51:
                mediaPlayer = MediaPlayer.create(this, R.raw.digit_3);
                mediaPlayer.start();
                break;
            case 52:
                mediaPlayer = MediaPlayer.create(this, R.raw.digit_4);
                mediaPlayer.start();
                break;
            case 53:
                mediaPlayer = MediaPlayer.create(this, R.raw.digit_5);
                mediaPlayer.start();
                break;
            case 54:
                mediaPlayer = MediaPlayer.create(this, R.raw.digit_6);
                mediaPlayer.start();
                break;
            case 55:
                mediaPlayer = MediaPlayer.create(this, R.raw.digit_7);
                mediaPlayer.start();
                break;
            case 56:
                mediaPlayer = MediaPlayer.create(this, R.raw.digit_8);
                mediaPlayer.start();
                break;
            case 57:
                mediaPlayer = MediaPlayer.create(this, R.raw.digit_9);
                mediaPlayer.start();
                break;

            case 97:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_a);
                mediaPlayer.start();
                break;
            case 98:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_b);
                mediaPlayer.start();
                break;
            case 99:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_c);
                mediaPlayer.start();
                break;
            case 100:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_d);
                mediaPlayer.start();
                break;
            case 101:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_e);
                mediaPlayer.start();
                break;
            case 102:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_f);
                mediaPlayer.start();
                break;
            case 103:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_g);
                mediaPlayer.start();
                break;
            case 104:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_h);
                mediaPlayer.start();
                break;
            case 105:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_i);
                mediaPlayer.start();
                break;
            case 106:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_j);
                mediaPlayer.start();
                break;
            case 107:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_k);
                mediaPlayer.start();
                break;
            case 108:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_l);
                mediaPlayer.start();
                break;
            case 109:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_m);
                mediaPlayer.start();
                break;
            case 110:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_n);
                mediaPlayer.start();
                break;
            case 111:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_o);
                mediaPlayer.start();
                break;
            case 112:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_p);
                mediaPlayer.start();
                break;
            case 113:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_qu);
                mediaPlayer.start();
                break;
            case 114:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_r);
                mediaPlayer.start();
                break;
            case 115:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_s);
                mediaPlayer.start();
                break;
            case 116:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_t);
                mediaPlayer.start();
                break;
            case 117:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_u);
                mediaPlayer.start();
                break;
            case 118:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_v);
                mediaPlayer.start();
                break;
            case 119:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_w);
                mediaPlayer.start();
                break;
            case 120:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_x);
                mediaPlayer.start();
                break;
            case 121:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_y);
                mediaPlayer.start();
                break;
            case 122:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_z);
                mediaPlayer.start();
                break;

            case 32:
                audioManager.playSoundEffect(AudioManager.FX_KEYPRESS_SPACEBAR);
                break;
            case Keyboard.KEYCODE_DONE:
            case 10:
                audioManager.playSoundEffect(AudioManager.FX_KEYPRESS_RETURN);
                break;
            case Keyboard.KEYCODE_DELETE:
                audioManager.playSoundEffect(AudioManager.FX_KEYPRESS_DELETE);
                break;
            default: audioManager.playSoundEffect(AudioManager.FX_KEYPRESS_STANDARD);
        }
    }

    private void playClickSw(int keyCode){
        AudioManager audioManager = (AudioManager) getSystemService(AUDIO_SERVICE);
        MediaPlayer mediaPlayer = null;
        switch(keyCode) {
            case 48:
                mediaPlayer = MediaPlayer.create(this, R.raw.digit_0);
                mediaPlayer.start();
                break;
            case 49:
                mediaPlayer = MediaPlayer.create(this, R.raw.digit_1);
                mediaPlayer.start();
                break;
            case 50:
                mediaPlayer = MediaPlayer.create(this, R.raw.digit_2);
                mediaPlayer.start();
                break;
            case 51:
                mediaPlayer = MediaPlayer.create(this, R.raw.digit_3);
                mediaPlayer.start();
                break;
            case 52:
                mediaPlayer = MediaPlayer.create(this, R.raw.digit_4);
                mediaPlayer.start();
                break;
            case 53:
                mediaPlayer = MediaPlayer.create(this, R.raw.digit_5);
                mediaPlayer.start();
                break;
            case 54:
                mediaPlayer = MediaPlayer.create(this, R.raw.digit_6);
                mediaPlayer.start();
                break;
            case 55:
                mediaPlayer = MediaPlayer.create(this, R.raw.digit_7);
                mediaPlayer.start();
                break;
            case 56:
                mediaPlayer = MediaPlayer.create(this, R.raw.digit_8);
                mediaPlayer.start();
                break;
            case 57:
                mediaPlayer = MediaPlayer.create(this, R.raw.digit_9);
                mediaPlayer.start();
                break;

            case 97:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_a);
                mediaPlayer.start();
                break;
            case 98:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_b);
                mediaPlayer.start();
                break;
            case 99:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_c);
                mediaPlayer.start();
                break;
            case 100:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_d);
                mediaPlayer.start();
                break;
            case 101:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_e);
                mediaPlayer.start();
                break;
            case 102:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_f);
                mediaPlayer.start();
                break;
            case 103:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_g);
                mediaPlayer.start();
                break;
            case 104:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_h);
                mediaPlayer.start();
                break;
            case 105:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_i);
                mediaPlayer.start();
                break;
            case 106:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_j);
                mediaPlayer.start();
                break;
            case 107:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_k);
                mediaPlayer.start();
                break;
            case 108:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_l);
                mediaPlayer.start();
                break;
            case 109:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_m);
                mediaPlayer.start();
                break;
            case 110:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_n);
                mediaPlayer.start();
                break;
            case 111:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_o);
                mediaPlayer.start();
                break;
            case 112:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_p);
                mediaPlayer.start();
                break;
            case 114:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_r);
                mediaPlayer.start();
                break;
            case 115:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_s);
                mediaPlayer.start();
                break;
            case 116:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_t);
                mediaPlayer.start();
                break;
            case 117:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_u);
                mediaPlayer.start();
                break;
            case 118:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_v);
                mediaPlayer.start();
                break;
            case 119:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_w);
                mediaPlayer.start();
                break;
            case 121:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_y);
                mediaPlayer.start();
                break;
            case 122:
                mediaPlayer = MediaPlayer.create(this, R.raw.letter_sound_z);
                mediaPlayer.start();
                break;

            case 32:
                audioManager.playSoundEffect(AudioManager.FX_KEYPRESS_SPACEBAR);
                break;
            case Keyboard.KEYCODE_DONE:
            case 10:
                audioManager.playSoundEffect(AudioManager.FX_KEYPRESS_RETURN);
                break;
            case Keyboard.KEYCODE_DELETE:
                audioManager.playSoundEffect(AudioManager.FX_KEYPRESS_DELETE);
                break;
            default: audioManager.playSoundEffect(AudioManager.FX_KEYPRESS_STANDARD);
        }
    }

    @Override
    public void onText(CharSequence charSequence) {

    }

    @Override
    public void swipeLeft() {

    }

    @Override
    public void swipeRight() {

    }

    @Override
    public void swipeDown() {

    }

    @Override
    public void swipeUp() {

    }
}

File: app/src/main/java/ai/elimu/keyboard/ImeService.kt
package ai.elimu.keyboard

import ai.elimu.keyboard.receiver.StudentUpdatedReceiver
import ai.elimu.keyboard.util.MediaPlayerHelper
import android.inputmethodservice.InputMethodService
import android.inputmethodservice.Keyboard
import android.inputmethodservice.KeyboardView
import android.inputmethodservice.KeyboardView.OnKeyboardActionListener
import android.media.AudioManager
import android.preference.PreferenceManager
import android.util.Log
import android.view.KeyEvent
import android.view.View

class ImeService : InputMethodService(), OnKeyboardActionListener {
    
    private val TAG = javaClass.getName()

    private var keyboardView: KeyboardView? = null

    private var keyboard: Keyboard? = null

    private var caps = false

    override fun onCreate() {
        Log.i(TAG, "onCreate")
        super.onCreate()
    }

    override fun onCreateInputView(): View {
        Log.i(TAG, "onCreateInputView")

        // Personalize available letters/numbers
        val sharedPreferences =
            PreferenceManager.getDefaultSharedPreferences(applicationContext)
        val availableLettersSet =
            sharedPreferences.getStringSet(StudentUpdatedReceiver.PREF_STUDENT_LETTERS, null)
        Log.d(TAG, "availableLettersSet: $availableLettersSet")
        val availableNumbersSet =
            sharedPreferences.getStringSet(StudentUpdatedReceiver.PREF_STUDENT_NUMBERS, null)
        Log.d(TAG, "availableNumbersSet: $availableNumbersSet")

        // TODO: add custom Keyboard
        if (availableLettersSet == null) {
            keyboard = Keyboard(this, R.xml.qwerty)
        } else if (availableLettersSet.size <= 3) {
            keyboard = Keyboard(this, R.xml.qwerty)
        } else if (availableLettersSet.size == 4) {
            keyboard = Keyboard(this, R.xml.qwerty_4)
        } else if (availableLettersSet.size == 5) {
            keyboard = Keyboard(this, R.xml.qwerty_5)
        } else if (availableLettersSet.size == 6) {
            keyboard = Keyboard(this, R.xml.qwerty_6)
        } else if (availableLettersSet.size == 7) {
            keyboard = Keyboard(this, R.xml.qwerty_7)
        } else if (availableLettersSet.size == 8) {
            keyboard = Keyboard(this, R.xml.qwerty_8)
        } else {
            keyboard = Keyboard(this, R.xml.qwerty_8)
        }

        keyboardView = layoutInflater.inflate(R.layout.keyboard, null) as KeyboardView
        keyboardView!!.setKeyboard(keyboard)
        keyboardView!!.isPreviewEnabled = false
        keyboardView!!.setOnKeyboardActionListener(this)

        return keyboardView!!
    }

    override fun updateFullscreenMode() {
        Log.i(TAG, "updateFullscreenMode")
        //        super.updateFullscreenMode();
    }

    override fun onPress(i: Int) {
    }

    override fun onRelease(i: Int) {
    }

    override fun onKey(primaryCode: Int, keyCodes: IntArray?) {
        val ic = getCurrentInputConnection()

        val locale = resources.configuration.locale
        Log.d(TAG, "locale: $locale")
        playClick(primaryCode)

        when (primaryCode) {
            Keyboard.KEYCODE_DELETE -> ic.deleteSurroundingText(1, 0)
            Keyboard.KEYCODE_SHIFT -> {
                caps = !caps
                keyboard!!.setShifted(caps)
                keyboardView!!.invalidateAllKeys()
            }

            Keyboard.KEYCODE_DONE -> ic.sendKeyEvent(
                KeyEvent(
                    KeyEvent.ACTION_DOWN,
                    KeyEvent.KEYCODE_ENTER
                )
            )

            else -> {
                var code = primaryCode.toChar()
                if (Character.isLetter(code) && caps) {
                    code = code.uppercaseChar()
                }
                ic.commitText(code.toString(), 1)
            }
        }
    }

    private fun playClick(keyCode: Int) {
        val audioManager = getSystemService(AUDIO_SERVICE) as AudioManager

        when (keyCode) {
            48 -> MediaPlayerHelper.play(applicationContext, R.raw.digit_0)
            49 -> MediaPlayerHelper.play(applicationContext, R.raw.digit_1)
            50 -> MediaPlayerHelper.play(applicationContext, R.raw.digit_2)
            51 -> MediaPlayerHelper.play(applicationContext, R.raw.digit_3)
            52 -> MediaPlayerHelper.play(applicationContext, R.raw.digit_4)
            53 -> MediaPlayerHelper.play(applicationContext, R.raw.digit_5)
            54 -> MediaPlayerHelper.play(applicationContext, R.raw.digit_6)
            55 -> MediaPlayerHelper.play(applicationContext, R.raw.digit_7)
            56 -> MediaPlayerHelper.play(applicationContext, R.raw.digit_8)
            57 -> MediaPlayerHelper.play(applicationContext, R.raw.digit_9)
            97 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_a)
            98 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_b)
            99 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_c)
            100 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_d)
            101 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_e)
            102 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_f)
            103 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_g)
            104 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_h)
            105 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_i)
            106 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_j)
            107 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_k)
            108 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_l)
            109 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_m)
            110 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_n)
            111 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_o)
            112 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_p)
            113 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_qu)
            114 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_r)
            115 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_s)
            116 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_t)
            117 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_u)
            118 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_v)
            119 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_w)
            120 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_x)
            121 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_y)
            122 -> MediaPlayerHelper.play(applicationContext, R.raw.letter_sound_z)
            32 -> audioManager.playSoundEffect(AudioManager.FX_KEYPRESS_SPACEBAR)
            Keyboard.KEYCODE_DONE, 10 -> audioManager.playSoundEffect(AudioManager.FX_KEYPRESS_RETURN)
            Keyboard.KEYCODE_DELETE -> audioManager.playSoundEffect(AudioManager.FX_KEYPRESS_DELETE)
            else -> audioManager.playSoundEffect(AudioManager.FX_KEYPRESS_STANDARD)
        }
    }

    override fun onText(charSequence: CharSequence?) {
    }

    override fun swipeLeft() {
    }

    override fun swipeRight() {
    }

    override fun swipeDown() {
    }

    override fun swipeUp() {
    }
}
